std/socket.d

```diff
@@ -3582,17 +3582,17 @@ class UdpSocket: Socket
             protected pure nothrow @safe Socket accepting() { assert(0); }
             @trusted Socket accept() { assert(0); }
             nothrow @nogc @trusted void shutdown(SocketShutdown how) { assert(0); }
-            nothrow @nogc @trusted void close() { assert(0); }
+            nothrow @nogc @trusted void close() scope { assert(0); }
             @property @trusted Address remoteAddress() { assert(0); }
             @property @trusted Address localAddress() { assert(0); }
-            @trusted ptrdiff_t send(const(void)[] buf, SocketFlags flags) { assert(0); }
-            @safe ptrdiff_t send(const(void)[] buf) { assert(0); }
+            @trusted ptrdiff_t send(const(void)[] buf, SocketFlags flags) scope { assert(0); }
+            @safe ptrdiff_t send(const(void)[] buf) scope { assert(0); }
             @trusted ptrdiff_t sendTo(const(void)[] buf, SocketFlags flags, Address to) { assert(0); }
             @safe ptrdiff_t sendTo(const(void)[] buf, Address to) { assert(0); }
             @trusted ptrdiff_t sendTo(const(void)[] buf, SocketFlags flags) { assert(0); }
             @safe ptrdiff_t sendTo(const(void)[] buf) { assert(0); }
-            @trusted ptrdiff_t receive(void[] buf, SocketFlags flags) { assert(0); }
-            @safe ptrdiff_t receive(void[] buf) { assert(0); }
+            @trusted ptrdiff_t receive(void[] buf, SocketFlags flags) scope { assert(0); }
+            @safe ptrdiff_t receive(void[] buf) scope { assert(0); }
```

Comment:
There seems to be a conflict of objectives:
To the best of my knowledge - with current DIP1000.md and it's implementation (scope is a part of (member) function's signature/mangling) - I don't see a way to have std.socket -dip1000 compilable without breaking user code (i.e. change class Socket). According to https://github.com/dlang/DIPs/blob/master/DIPs/DIP1000.md#breaking-changes--deprecation-process it's known there will be breaking changes; IMHO this PR is an example for coming breakages, but I now understand - supposed my PR is correct - it shouldn't be merged without a deprecation process in concert with -transition=safe.

Reply:
CC @andralex @MartinNowak @WalterBright 
What's the strategy for such errors resulting from DIP1000 upgrades?
Do we need to introduce some `@__future` like behavior in the compiler to issue deprecations?

Reply:
It simply means that Socket classes and sub-classes cannot be used as scoped classes with DIP1000, because it's unclear whether they escape sth. or not. It should remain fully useable with GC allocated Socket classes though.
As @CyberShadow said, if Socket is not final we cannot suddenly require all derivatives to obey to not escaping sth.
That's unfortunate but a big restriction with open interface&classes in APIs.
I asked for `@future` to become useful for such deprecations as well, but it made it to a mentioning in https://github.com/dlang/DIPs/blob/master/DIPs/DIP1007.md#user-content-proposal under **6.**.
I don't think exempting Socket from DIP1000 is a big issue, the whole module is outdated (by todays D standards) and class based sockets are questionable. Here is a possible replacement https://github.com/MartinNowak/io (though I still need to get `return scope` correct).

Reply:
There is no two-tier memory safety.
D can't ever claim memory safety of @safe code, if it starts to exempt some phobos module(s) from memory checks which possibly may then escape pointers to expired stack frames or alike.
Walter said at DConf2017, he believes there is a tsunami coming and predicted, it will kill the C language. That's my opinion too ref. growing importance of memory safety for language selection.
And D still isn't ahead of the caravan. https://www.youtube.com/watch?v=Lo6Q2vB9AAg

Was there a survey, whether users put up with code breakage mitigated by a reasonable deprecation duration, when they get memory safety in return?

Reply:
> D can't ever claim memory safety of @safe code, if it starts to exempt some phobos module(s) from memory checks which possibly may then escape pointers to expired stack frames or alike.

We can exempt Socket from being RC or stack-allocatable though. It just requires GC to be memory-safe.

Reply:
> Walter said at DConf2017, he believes there is a tsunami coming and predicted, it will kill the C language. That's my opinion too ref. growing importance of memory safety for language selection.
> And D still isn't ahead of the caravan.

I've recently posted a status update and am actively working (quite a lot) on an upcoming proposal and prototype implementation.
https://forum.dlang.org/post/p6cq7f$2cg$1@digitalmars.com

Reply:
I take this as PR rejection and will close it

Reply:
How about simply making `std.socket` @system so it will compile with -dip1000 and move on?

Reply:
Yes!! It's as simple as that to avoid changing class socket and be -dip1000 compilable

Reply:
Let's do it, then.